src/Microsoft.DotNet.Interactive.HttpRequestParser/HttpSyntaxNodeOrToken.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.DotNet.Interactive.HttpRequest;
+
+internal abstract class HttpSyntaxNodeOrToken
+{
+    private protected HttpSyntaxNodeOrToken(SourceText sourceText, HttpSyntaxTree? syntaxTree)
+    {
+        SourceText = sourceText;
+        SyntaxTree = syntaxTree;
+    }
+
+    protected SourceText SourceText { get; }
+
+    public HttpSyntaxNode? Parent { get; internal set; }
+
+    public abstract TextSpan Span { get; }
```

Comment:
From our discussions yesterday, I think it may be useful to have a `IsMissing` property (that returns `true` if the span has `0` length. I think we'll need it pretty quickly once we start consuming the parser - but we can probably wait and add it later when we first need it...

Reply:
+1 to adding this only when we decide we need it.

Reply:
One other thing I realized is that Roslyn has both `Span` and `FullSpan` (in parallel to `Text` and `FullText`). `Span` excludes (leading and trailing) trivia while `FullSpan` does not.

I think we may also end up needing similar APIs as otherwise it can get cumbersome to walk the tree and figure out what the first and last significant token is each time...

Reply:
Makes sense.

Reply:
This was addressed in #3122